mango-core/src/main/scala/org/bdgenomics/mango/io/VcfReader.scala

```diff
@@ -0,0 +1,249 @@
+
+  def suffixes = Array(".vcf", ".vcf.gz")
+
+  /**
+   * Loads file from local filesystem, http or hdfs.
+   *
+   * @param fp url to file
+   * @param regions Iterable of ReferenceRegions
+   * @param sc Optional SparkContext, only used for HDFS files
+   * @return Array of Specific Records
+   */
+  def loadDataAndSamplesFromSource(fp: String, regions: Option[Iterable[ReferenceRegion]], sc: Option[SparkContext] = None): Tuple2[Seq[Sample], Array[ADAMVariantContext]] = {
+
+    if (new java.io.File(fp).exists) { // if fp exists in local filesystem, load locally
+      val results = load(fp, regions, true)
+
+      val genotypeSamples = results._1.getGenotypeSamples.map(s => {
+        Sample.newBuilder()
+          .setName(s)
+          .setId(s)
+          .build()
+      })
+
+      (genotypeSamples, results._2)
+
+    } else if (fp.startsWith("http")) { // TODO check correctly
+      // http file
+      val results = load(fp, regions, false)
+
+      val genotypeSamples = results._1.getGenotypeSamples.map(s => {
+        Sample.newBuilder()
+          .setName(s)
+          .setId(s)
+          .build()
+      })
+
+      (genotypeSamples, results._2)
+
+    } else {
+      require(sc.isDefined, "HDFS requires a SparkContext to run")
+      val vcDataset = loadHDFS(sc.get, fp, regions)
+      (vcDataset._1.samples, vcDataset._2)
+    }
+  }
+
+  /**
+   * *
+   * Loads records from vcf or compressed vcf files overlapping queried reference regions.
+   *
+   * @param fp filepath
+   * @param regionsOpt Option of iterator of ReferenceRegions to query
+   * @param local Boolean specifying whether file is local or remote.
+   * @return Iterator of VariantContext
+   */
+  def load(fp: String, regionsOpt: Option[Iterable[ReferenceRegion]], local: Boolean): Tuple2[VCFHeader, Array[ADAMVariantContext]] = {
+
+    val codec = new VCFCodec()
+
+    // if not valid throw Exception
+    if (!isValidSuffix(fp)) {
+      invalidFileException(fp)
+    }
+
+    val isGzipped = fp.endsWith(".gz")
+
+    if (local) {
+      // create index file
+      createIndex(fp, codec)
+    }
+
+    val vcfReader: AbstractFeatureReader[HtsjdkVariantContext, LineIterator] =
+      AbstractFeatureReader.getFeatureReader(fp, codec, true)
+
+    val header: VCFHeader = vcfReader.getHeader().asInstanceOf[VCFHeader]
+
+    val results =
+      if (regionsOpt.isDefined) {
+        val regions = regionsOpt.get
+
+        val queries =
+          try {
+            val dictionary = vcfReader.getSequenceNames()
+
+            // modify chr prefix, if this file uses chr prefixes.
+            val hasChrPrefix = dictionary.head.startsWith("chr")
+
+            regions.map(r => {
+              LazyMaterialization.modifyChrPrefix(r, hasChrPrefix)
+            })
+          } catch {
+            case e: NullPointerException => {
+              log.warn(s"File ${fp} did not have sequence header")
+              regions
+            }
+          }
+
+        if (isGzipped)
```

Comment:
Does a distinction between GZIP and block-compressed GZIP (BGZF) need to be made here?

Reply:
not sure, do you have a BGZF reference I can play with?

Reply:
ADAM `adam-core/src/test/resources` has

```
test.compressed.bcf
test.uncompressed.bcf
test.vcf
test.vcf.bgz
test.vcf.bgzf.gz
test.vcf.gz
```

Disq `src/test/resources` has

```
HiSeq.10000.vcf.bgz
HiSeq.10000.vcf.bgz.tbi
HiSeq.10000.vcf.bgzf.gz
HiSeq.10000.vcf.bgzf.gz.tbi
test.vcf
test.vcf.bgz
test.vcf.bgzf.gz
test.vcf.gz
```

Reply:
Awesome, thanks!

Reply:
Thanks for the thoughts @heuermh I pushed some tests and a fix.